index.html

```diff
@@ -140,67 +166,250 @@ <h2>
+        badge indicators from documents and service workers, respectively.
+      </p>
       <pre class="idl">
-        [Exposed=(Window)]
-        namespace Badge {
-          void set(optional [EnforceRange] unsigned long long contents);
-          void clear();
+        // Methods only exposed on documents.
+        [SecureContext]
+        partial interface Navigator {
+          Promise&lt;void&gt; setClientBadge(optional [EnforceRange] unsigned long long contents);
+          Promise&lt;void&gt; clearClientBadge();
+        };
+
+        // Methods exposed on both documents and service workers.
+        [SecureContext]
+        interface mixin NavigatorBadge {
+          Promise&lt;void&gt; setAppBadge(optional [EnforceRange] unsigned long long contents);
+          Promise&lt;void&gt; clearAppBadge();
         };
+
+        Navigator includes NavigatorBadge;
+        WorkerNavigator includes NavigatorBadge;
       </pre>
       <p>
-        The <dfn>Badge</dfn> namespace provides functions for setting and
-        clearing the application badge indicator.
+        User agents that never display document badges SHOULD NOT expose the
+        {{Navigator/setClientBadge}} and {{Navigator/clearClientBadge}}
+        methods. Similarly, user agents that never display application badges
+        SHOULD NOT expose the {{NavigatorBadge/setAppBadge}} and
+        {{NavigatorBadge/clearAppBadge}} methods.
+      </p>
+      <p class="note">
+        This is important as a feature-detection mechanism to allow sites to
+        determine whether setting a badge will have any effect, based on the
+        presence of the methods. In particular, sites can use the absence of
+        {{Navigator/setClientBadge}} as a condition for displaying a fallback
+        badge in the page's <a data-cite=
+        "html#the-title-attribute"><code>title</code></a> or <a data-cite=
+        "html#rel-icon"><code>icon</code></a>.
+      </p>
+      <p class="note">
+        These methods return a {{Promise}}, even though there are no
+        asynchronous failure conditions (the only failure conditions are WebIDL
+        type errors). This is to allow us to add asynchronous failure
+        conditions in the future without altering how type errors are returned.
       </p>
       <section>
         <h3>
-          <dfn>set()</dfn> function
+          <dfn data-dfn-for="Navigator">setClientBadge()</dfn> method
         </h3>
         <p>
-          When the <code>set()</code> function is called with argument
+          When the <code>setClientBadge()</code> method is called with argument
           <var>contents</var>, run the following steps:
         </p>
         <ol>
-          <li>If <var>contents</var> is omitted, set the application's badge to
-          <a>flag</a>.
+          <li>Let <var>document</var> be the <a data-cite=
+          "html#current-settings-object">current settings object</a>'s
```

Comment:
I see what you mean about this being kind of roundabout. Is there a reason we can't use the current execution context instead of the current document?

Reply:
**Note:** We already depend on the current execution context in the matching applications algorithm

Reply:
This seems to be "the way" to do it.